Don 880 prb inside payment element

bdsl opened the pull request.

bdsl commented on src/app/stripe.service.ts:
```diff
@@ -86,71 +84,6 @@ export class StripeService {
     this.lastCardCountry = cardCountry;
   }
 
-  async confirmPaymentWithNewCardOrPRB(
```
this function was unused before this PR - should have been deleted then.

bdsl commented on src/app/donation-start/donation-start-form/donation-start-form.component.ts:
```diff
@@ -1185,7 +1171,7 @@ export class DonationStartFormComponent implements AfterContentChecked, AfterCon
 
     this.currencySymbol = getCurrencySymbol(this.campaign.currencyCode, 'narrow', 'en-GB');
 
-    if (environment.psps.stripe.enabled && this.campaign.charity.stripeAccountId) {
```
It would still have a code change to set `enabled` to false and a new deployment wouldn't it? So I can put this back how it was if you'd like but I'm not sure to be honest it saves us much time in the event a stripe outage. And having it adds a bit of complexity when we're reading the code.

NoelLH replied:
@bdsl I think that having flags for this kind of emergency change in big campaigns is still worth it to avoid having to do anything more complicated while under time pressure.

bdsl replied:
OK, I've put the flag back in bdb6cb05

bdsl commented on src/app/donation-start/donation-start-form/donation-start-form.component.ts:
```diff
@@ -1330,79 +1316,6 @@ export class DonationStartFormComponent implements AfterContentChecked, AfterCon
-      throw new Error("Stripe elements not ready");
-    }
-
-    this.paymentRequestButton = this.stripeService.getPaymentRequestButton(donation, this.stripeElements, paymentRequestResultObserver);
```
:+1: . It would be great to have a static analysis check to say we're don't have dead code, unless either we've stated that its there for some purpose that the SA tool can't detect, like use by code we intend to write in future, or dynamic use.

Done in 0fbe33dd

bdsl commented on src/app/donation-start/donation-start-form/donation-start-form.component.ts:
```diff
@@ -76,13 +75,11 @@ export class DonationStartFormComponent implements AfterContentChecked, AfterCon
   cardHandler = this.onStripeCardChange.bind(this);
-  paymentRequestButton: StripePaymentRequestButtonElement | null;
 
   requestButtonShown = false;
```
Thanks, done in [444633b](https://github.com/thebiggive/donate-frontend/pull/1295/commits/444633be2be41ceadd20b6d4c90e62ed1baba998) (along with some other stuff I noticed was dead)

NoelLH commented on src/app/stripe.service.ts:
```diff
@@ -86,71 +81,6 @@ export class StripeService {
-          };
-        }
-
-        let thePaymentMethodEvent = this.paymentMethodEvents.get(donation.donationId);
```
I think `paymentMethodEvents` is redundant now too.

Probably `paymentMethodIds` and the code managing them as well?

bdsl replied:
You're right: 79f3c405

NoelLH commented:
Thanks for the updates @bdsl.

Sorry for reviewing in bits, but looking at the updated files again I think there's slightly more we can still tidy and in particular we are probably missing the update for Matomo events I mentioned on Jira:
* `canUsePaymentRequest()` and the var it uses can probably go – I certainly agree static analysis for dead code would be handy here if you have any suggestions
* I don't think we have any reason to set `handleActions` on [confirmCardPayment()](https://stripe.com/docs/js/payment_intents/confirm_card_payment) now and can probably delete the options key and use the default (true)
* connected to the above, we probably need to use Stripe.js events to pick up on whether the payment is being made with a wallet or not, in order for custom events to remain consistent with previous data and give us the info to help spot any edge case problems with donors using wallets

bdsl commented:
> Thanks for the updates @bdsl.
> 
> Sorry for reviewing in bits, but looking at the updated files again I think there's slightly more we can still tidy and in particular we are probably missing the update for Matomo events I mentioned on Jira:

Thanks. I've removed canUsePaymentRequest, and handleActions. and added tracking for successful payments that should indicate whether its with a card or a PRB. https://stripe.com/docs/js/element/events/on_change?type=paymentElement doesn't really document the string they use to indicate PRBs - it just gives 'card' as an example value, so that line of code might need some adjustment once this is tested on staging with Google pay and Apple Pay.

NoelLH commented on src/app/donation-start/donation-start-form/donation-start-form.component.ts:
```diff
@@ -637,8 +626,11 @@ export class DonationStartFormComponent implements AfterContentChecked, AfterCon
+   * a value key as expected here. I'm not sure why that isn't included in the TS StripeElementChangeEvent interface.
+   */
+  async onStripeCardChange(state: StripeElementChangeEvent & {value: {type: string}}) {
     this.addStripeCardBillingValidators();
```
```suggestion
    console.log('Change evt', state);
    this.addStripeCardBillingValidators();
```

bdsl commented:
Re static analysis for dead code - I had a brief go using https://github.com/webpro/knip yesterday, but it found so many false positives it looked like it would need a lot of config to be useful, and I couldn't find any ready-made way to integrate it with an Angular app.

We could probably get something simpler to detect unused exports, but that probably wouldn't help us find unused elements within e.g. a class that is used, e.g. https://www.npmjs.com/package/ts-unused-exports
